scylla/src/transport/connection.rs

```diff
@@ -821,33 +829,52 @@ impl Connection {
         }
     }
 
+    fn alloc_stream_id(
+        handler_map: &StdMutex<ResponseHandlerMap>,
+        response_handler: ResponseHandler,
+    ) -> Option<i16> {
+        // We are guaranteed here that handler_map will not be locked
+        // by anybody else, so we can do try_lock().unwrap()
+        let mut lock = handler_map.try_lock().unwrap();
+        if let Some(stream_id) = lock.allocate(response_handler) {
+            Some(stream_id)
+        } else {
+            // TODO: Handle this error better, for now we drop this
+            // request and return an error to the receiver
+            error!("Could not allocate stream id");
+            None
+        }
+    }
+
     async fn writer(
         mut write_half: (impl AsyncWrite + Unpin),
         handler_map: &StdMutex<ResponseHandlerMap>,
         mut task_receiver: mpsc::Receiver<Task>,
     ) -> Result<(), QueryError> {
+        const OUTPUT_BUFFER_MAX: usize = 8192;
         // When the Connection object is dropped, the sender half
         // of the channel will be dropped, this task will return an error
         // and the whole worker will be stopped
-        while let Some(task) = task_receiver.recv().await {
-            let stream_id = {
-                // We are guaranteed here that handler_map will not be locked
-                // by anybody else, so we can do try_lock().unwrap()
-                let mut lock = handler_map.try_lock().unwrap();
-
-                if let Some(stream_id) = lock.allocate(task.response_handler) {
-                    stream_id
-                } else {
-                    // TODO: Handle this error better, for now we drop this
-                    // request and return an error to the receiver
-                    error!("Could not allocate stream id");
-                    continue;
+        while let Some(mut task) = task_receiver.recv().await {
+            let mut num_requests = 0;
+            let mut total_sent = 0;
+            while let Some(stream_id) = Self::alloc_stream_id(handler_map, task.response_handler) {
+                let mut req = task.serialized_request;
+                req.set_stream(stream_id);
+                let req_data: &[u8] = req.get_data();
+                total_sent += req_data.len();
+                num_requests += 1;
+                write_half.write_all(req_data).await?;
+                if total_sent >= OUTPUT_BUFFER_MAX {
+                    break;
                 }
-            };
-
-            let mut req = task.serialized_request;
-            req.set_stream(stream_id);
-            write_half.write_all(req.get_data()).await?;
+                task = match task_receiver.try_recv() {
+                    Ok(t) => t,
+                    Err(_) => break,
+                }
+            }
+            debug!("Sending {} requests; {} bytes", num_requests, total_sent);
```

Comment:
I'm not sure if this message is more appropriate for the `debug` or `trace` level... Could you explain the choice? Personally, I'm never sure which one to choose when I write own code, but I have a feeling that `trace` should be reserved for high-volume, low-level debugging information - and this message will be printed a lot, especially if there are lots of requests being made.